Test: src/graphql/types/Community/updater.ts

NishantSinghhhhh opened the pull request:
What kind of change does this PR introduce?
Adding tests for updater.ts

Issue Number:
Fixes: #3085 

Snapshots/Videos:
[Screencast from 2025-02-01 04-25-01.webm](https://github.com/user-attachments/assets/7d6d7f2b-93ef-410e-b185-be4d0af61d71)


If relevant, did you update the documentation?
N/A

Summary
This PR improves the updater.ts file by:

Replacing any types with more specific TypeScript types to improve type safety.
Added tests to verify the updater.ts logic, including database error handling, user retrieval, and authentication/authorization checks.

Does this PR introduce a breaking change?
No

Checklist
CodeRabbit AI Review
 I have reviewed and addressed all critical issues flagged by CodeRabbit AI
 I have implemented or provided justification for each non-critical suggestion
 I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented
Test Coverage
 I have written tests for all new changes/features
 I have verified that test coverage meets or exceeds 95%
 I have run the test suite locally and all tests pass
Other Information
<!-- Add extra information about this PR here -->
Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?
Yes ✅

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced community update functionality that validates user authentication and returns relevant user details when available, improving error handling for unauthorized access.
  - Introduced a mock logger for testing purposes, enhancing logging capabilities in the testing environment.
  - Added UUID generation capabilities to the project.

- **Tests**
  - Added a comprehensive test suite for the community resolver's updater method, covering various scenarios and ensuring correct functionality and error handling.
  - Implemented a mock logger to streamline testing processes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

NishantSinghhhhh commented:
@palisadoes , there are type errors coming from files that i did not even touch

palisadoes commented:
Did you merge with the latest upstream? 


The sign up file was fixed a few days ago

NishantSinghhhhh commented:
> Did you merge with the latest upstream?
> 
> The sign up file was fixed a few days ago

Yes I did , My code was up to date

NishantSinghhhhh commented:
@palisadoes , I will be adding more tests in this PR , Do not merge it now , there aree lot many test cases that are left to be covered

NishantSinghhhhh commented:
will be resolving rabbits error once , all my tests have been added

NishantSinghhhhh commented:
@coderabbitai full review

palisadoes commented:
Please make coderabbit approve your PR

NishantSinghhhhh commented on test/utilities/mockLogger.ts:
```diff
@@ -0,0 +1,28 @@
+import { vi } from "vitest";
+
+export const createMockLogger = (): FastifyBaseLogger => {
+	const logger = {
+		error: vi.fn(),
+		warn: vi.fn(),
+		info: vi.fn(),
+		debug: vi.fn(),
+		trace: vi.fn(),
+		fatal: vi.fn(),
+		silent: vi.fn(),
+		child: () => createMockLogger(),
```
causing error

NishantSinghhhhh commented:
@palisadoes , there is 1 type errror that was coming in this PR related to uuid , for that I have installed a new dependency 
as the error was persistent

![Screenshot from 2025-02-02 13-48-09](https://github.com/user-attachments/assets/f5e2ca2c-efa3-4af0-a1e0-ee95ecffc845)

NishantSinghhhhh commented:
@coderabbitai full review

NishantSinghhhhh commented on test/graphql/types/Community/updater.test.ts:
```diff
@@ -0,0 +1,302 @@
+		});
+	});
+
+	it("should log warning and throw error when updater is not found", async () => {
+		ctx.drizzleClient.query.usersTable.findFirst.mockResolvedValue(undefined);
+
+		const testCommunity = {
+			...mockCommunity,
+			updaterId: "non-existent-id",
+		};
+
+		await expect(
+			CommunityResolver.updater(testCommunity, {}, ctx),
+		).rejects.toThrow(
+			new TalawaGraphQLError({
+				message: "Updater user not found",
+				extensions: {
+					code: "arguments_associated_resources_not_found",
+					issues: [{ argumentPath: ["updaterId"] }],
+				},
+			}),
+		);
+
+		expect(ctx.log.warn).toHaveBeenCalledWith(
+			`No user found for updaterId: ${testCommunity.updaterId}`,
+		);
+	});
```
This test case is failing [Type '"bad_user_input"' is not assignable to type '"arguments_associated_resources_not_found" | "forbidden_action" | "forbidden_action_on_arguments_associated_resources" | "unauthenticated" | "invalid_arguments" | "unauthorized_action" | "unauthorized_action_on_arguments_associated_resources" | "unauthorized_arguments" | "unexpected"'.ts(2322)
TalawaGraphQLError.ts(33, 2): The expected type comes from property 'code' which is declared here on type 'Maybe<GraphQLErrorExtensions> & TalawaGraphQLErrorExtensions'
(property) code: "bad_user_input"
No quick fixes available] , this makes the code quality bad i guess

NishantSinghhhhh commented on test/graphql/types/Community/updater.test.ts:
```diff
@@ -0,0 +1,302 @@
+		);
+	});
+
+	it("should correctly query the database for current user and updater user", async () => {
+		const updaterUser = {
+			id: "456",
+			name: "Jane Updater",
+			role: "user",
+			createdAt: new Date(),
+			updatedAt: null,
+		};
+
+		ctx.drizzleClient.query.usersTable.findFirst
+			.mockResolvedValueOnce(mockUser) // First call for current user
+			.mockResolvedValueOnce(updaterUser); // Second call for updater user
+
+		await CommunityResolver.updater(mockCommunity, {}, ctx);
+
+		expect(ctx.drizzleClient.query.usersTable.findFirst).toHaveBeenCalledTimes(
+			2,
+		);
+	});
```
done

NishantSinghhhhh commented:
@palisadoes , CodeRabbit is asking me to change the backend code , as it states that adding role based authorization , which will require changing a lot many backend files

NishantSinghhhhh commented:
@coderabbitai full review

NishantSinghhhhh commented:
@coderabbitai full review

NishantSinghhhhh commented:
@palisadoes, I have completed this PR and added as many test cases as possible. I also included the `uuid` dependency, as its absence was causing a type error in one of the files.

![Screenshot from 2025-02-02 13-48-09](https://github.com/user-attachments/assets/bd111b90-8700-4813-8dc1-c83d5e29d76b)


 Additionally, I have incorporated most of the genuine changes requested by **coderabbitai**.  

If there are any other important modifications you'd like me to make, please let me know, and I'll address them right away. Otherwise, I'll proceed with working on the next component. 🚀

NishantSinghhhhh commented:
@coderabbitai full review

NishantSinghhhhh commented:
@palisadoes , done

NishantSinghhhhh commented:
@palisadoes, I have completed this PR and added as many test cases as possible. Can you tell me what more things I can add to make it more mergeable
